File: app/src/main/java/com/song/deviceinfo/ui/hardware/HardwareFragment.java
package com.song.deviceinfo.ui.hardware;

import com.song.deviceinfo.ui.base.BaseAdapter;
import com.song.deviceinfo.ui.base.BaseViewModel;
import com.song.deviceinfo.ui.base.NormalAdapter;
import com.song.deviceinfo.ui.base.NormalFragment;

import java.util.List;

import androidx.core.util.Pair;
import androidx.lifecycle.ViewModelProviders;

/**
 * Created by chensongsong on 2021/9/29.
 */
public class HardwareFragment extends NormalFragment {

    @Override
    protected BaseAdapter crateAdapter() {
        return new NormalAdapter(getContext());
    }

    @Override
    protected BaseViewModel crateViewModel() {
        return ViewModelProviders.of(this).get(HardwareViewModel.class);
    }

    @Override
    protected List<Pair<String, String>> getNormalInfo() {
        return ((HardwareViewModel) viewModel).getHardwareInfo(getContext());
    }
}

File: app/src/main/java/com/song/deviceinfo/ui/hardware/HardwareFragment.kt
package com.song.deviceinfo.ui.hardware

import androidx.core.util.Pair
import androidx.lifecycle.ViewModelProviders
import com.song.deviceinfo.ui.base.BaseAdapter
import com.song.deviceinfo.ui.base.BaseViewModel
import com.song.deviceinfo.ui.base.NormalAdapter
import com.song.deviceinfo.ui.base.NormalFragment

/**
 * Created by chensongsong on 2021/9/29.
 */
class HardwareFragment : NormalFragment() {
    override fun createAdapter(): BaseAdapter<Pair<String, String>, *> {
        return NormalAdapter(requireContext())
    }
    
    override fun createViewModel(): BaseViewModel<Pair<String, String>> {
        return ViewModelProviders.of(this).get(HardwareViewModel::class.java)
    }
    
    override val normalInfo: List<Pair<String, String>>
        get() = (viewModel as HardwareViewModel).getHardwareInfo(requireContext())
}
